Improved Rational for on-chain size

kozross opened the pull request:
Designed to close #4250. I've also taken the liberty of moving `quotRem`, `divMod` and `abs` into `Numeric`, as they're not specific to rational numbers in any way. A full list of changes is in the changelog.

As evidence of its efficacy, here are the results from [our implementation](https://github.com/Liqwid-Labs/plutus-extra/pull/121), using the size testing approach [from here](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-size-check).

```
Rational (us versus Plutus)
  Eq
    On-chain size comparison for ==:                  OK
      72B vs 75B 
      ~4% smaller
    On-chain size comparison for /=:                  OK
      80B vs 91B 
      ~12% smaller
  Ord
    On-chain size comparison for compare:             OK
      109B vs 129B 
      ~16% smaller
    On-chain size comparison for <=:                  OK
      62B vs 64B 
      ~3% smaller
    On-chain size comparison for >=:                  OK
      62B vs 112B 
      ~45% smaller
    On-chain size comparison for <:                   OK
      62B vs 112B 
      ~45% smaller
    On-chain size comparison for >:                   OK
      62B vs 112B 
      ~45% smaller
    On-chain size comparison for min:                 OK
      70B vs 72B 
      ~3% smaller
    On-chain size comparison for max:                 OK
      70B vs 72B 
      ~3% smaller
  AdditiveGroup
    On-chain size comparison for +:                   OK
      159B vs 222B 
      ~28% smaller
    On-chain size comparison for zero:                OK
      24B vs 28B 
      ~14% smaller
    On-chain size comparison for -:                   OK
      159B vs 222B 
      ~28% smaller
    On-chain size comparison for negate:              OK
      295B (~0.5KiB) vs 326B (~0.5KiB) 
      ~10% smaller
    On-chain size comparison for negate (overloaded): OK
      35B vs 326B (~0.5KiB) 
      ~89% smaller
  MultiplicativeMonoid
    On-chain size comparison for *:                   OK
      151B vs 214B 
      ~29% smaller
    On-chain size comparison for one:                 OK
      24B vs 28B 
      ~14% smaller
  Other
    On-chain size comparison for %:                   OK
      163B vs 269B (~0.5KiB) 
      ~39% smaller
    On-chain size comparison for fromInteger:         OK
      24B vs 26B 
      ~8% smaller
    On-chain size comparison for numerator:           OK
      26B vs 28B 
      ~7% smaller
    On-chain size comparison for denominator:         OK
      26B vs 28B 
      ~7% smaller
    On-chain size comparison for round:               OK
      342B (~0.5KiB) vs 711B (~0.5KiB) 
      ~52% smaller
    On-chain size comparison for truncate:            OK
      30B vs 32B 
      ~6% smaller
    On-chain size comparison for properFraction:      OK
      57B vs 59B 
      ~3% smaller
    On-chain size comparison for recip:               OK
      92B vs 275B (~0.5KiB) 
      ~67% smaller
    On-chain size comparison for abs:                 OK
      644B (~0.5KiB) vs 767B (~0.5KiB) 
      ~16% smaller
    On-chain size comparison for abs (overloaded):    OK
      69B vs 767B (~0.5KiB) 
      ~91% smaller
```

michaelpj commented on plutus-tx/src/PlutusTx/Numeric.hs:
```diff
@@ -126,3 +146,33 @@ instance MultiplicativeSemigroup a => Semigroup (Product a) where
+-- builtins 'divideInteger' and 'modInteger' are implemented as the Haskell
+-- 'div' and 'mod' operations, which in turn are impemented as fst . divMod and
+-- snd . divMod, so we're calling Haskell's 'divMod' twice and throwing away
+-- useful information.  We could fix his by exposing 'divMod' as a PLC builtin,
```
I don't actually believe this is the case. `div` and `mod` typically have separate implementations which are put together to form `divMod`, and a quick look at the Core verifies that GHC is indeed specializing the implementation of the builtin functions to just call the underlying `div` or `mod`.

kozross replied:
This is a direct copy of the doc notes that already existed on this function.

michaelpj replied:
So it is. They're wrong!

kwxm replied:
>  div and mod typically have separate implementations which are put together to form divMod

@michaelpj I believe that I wrote the note being discussed here, which says that `div` and `mod` are derived from `divMod`.  That's based on [this code](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.Real.html#divMod) from `GHC,Real`.  Could you clarify what you're saying about the Core?  My point was that here we're implementing `divMod` as  `divMod x y = ( x `divideInteger` y, x `modInteger` y)` and the two functions being called on the right hand side will be calling Haskell's `divMod` function with the same arguments.

michaelpj replied:
The Haskell implementation of `divideInteger` uses the Haskell function `div`. GHC is smart enough to compile `div` into a division primitive, even though it is initially defined in terms of `divMod`. You can check the Core to see. So there is no extra overhead.

kwxm replied:
> The Haskell implementation of `divideInteger` uses the Haskell function `div`.

Yes, but my point is that the Haskell function `div` is defined in terms of the Haskell function `divMod`. It's effectively `fst . divMod` and this is inherited from gmp: gmp only has one basic integer division function which calculates the quotient and remainder simultaneously, because there's no quick way to do one without the other.  The comment above refers to the _Plutus_ function `divMod`, which is defined to be `(x 'divideInteger' y, x 'modInteger' y)`: in Plutus Core that will currently compile to a Scott encoded pair which calls `divideInteger` and `modInteger`, and both of these will call _Haskell's_ `divMod`, doing twice as much work as necessary.  If we had `divMod` as a PLC builtin then we could avoid this, but that would have been difficult without built-in pairs.  The statement that

> div and mod typically have separate implementations which are put together to form divMod

is false for both Haskell and GMP: it's  the other way around.

Anyway, none of this has anything to do with this PR!  It's a comment that was written a long time ago and has become slightly outdated with the advent of built-in pairs.

michaelpj commented on plutus-tx/CHANGELOG.md:
```diff
@@ -0,0 +1,21 @@
+
+### Removed
+
+* `Ratio` data structure.
```
This feels a bit needlessly breaking, although I checked and at least all uses in `plutus-apps` are `Ratio Integer`. I think I'm *probably* fine about it.

kozross replied:
The current design is creating on-chain space usage for no reason. Nothing _other_ than `Integer` can currently go into it.

michaelpj replied:
Other things *could* go in it. Nothing currently does. Anyway, I'm fine with removing it.

michaelpj commented on plutus-tx/CHANGELOG.md:
```diff
@@ -0,0 +1,21 @@
+### Removed
+
+* `Ratio` data structure.
+* `gcd` and `reduce` from `PlutusTx.Ratio`.
```
Why? I can see that you don't need them since your comparison and equality functions work regardless of whether they're reduced, but you might as well leave the functions.

kozross replied:
Good point. I can restore them. Notably, the comparison and equality functions that already exist in `plutus-tx` perform their reductions internally.

michaelpj commented on plutus-tx/plutus-tx.cabal:
```diff
@@ -1,6 +1,6 @@
 cabal-version: 2.2
 name: plutus-tx
-version: 0.1.0.0
+version: 0.2.0.0
```
We have no versioning policy currently, so there's no point doing this.

kozross replied:
As per my comment above vis a vis the changelogging, this would immensely help at _least_ one organization currently relying on Plutus. Upgrades are _really_ difficult otherwise at the moment.

michaelpj replied:
Literally no Cardano library other than the node has a version. And so long as people are picking things out by git revisions, having a version bound doesn't actually help you much.

michaelpj commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,337 @@
 import PlutusTx.Bool qualified as P
 import PlutusTx.Eq qualified as P
-import PlutusTx.ErrorCodes qualified as P
+-- import PlutusTx.ErrorCodes qualified as P
```
commented out code, and below

michaelpj reviewed (commented):
I think something like this is probably fine.

I do think this would have benefited from following the [new section in CONTRIBUTING](https://github.com/input-output-hk/plutus/blob/master/CONTRIBUTING.adoc#external-contributors) (only added recently, so no worries). There are a few controversial decisions here that it would be good to have discussed up front.

michaelpj commented:
Also, while I do like your size benchmarking suite, I still think it has to deal with the points I raised here: https://github.com/input-output-hk/plutus/issues/4250#issuecomment-981474284

So in particular: does this make your actual validator scripts noticeably smaller? If you're pulling in the overhead elsewhere then it may not save you much.

kozross commented on plutus-tx/CHANGELOG.md:
```diff
@@ -0,0 +1,21 @@
+# Revision history for `plutus-tx`
```
... why not start? Given the extreme pain and complexity Plutus upgrades have caused us (meaning MLabs) in the past, we'd certainly welcome this kind of tracking, and I'm sure we're not alone here.

michaelpj replied:
Sure, and we might do, but until we do, this is fairly pointless.

kozross commented:
I am happy to open issues to discuss the controversial decisions: which ones specifically? The PR is marked WIP for (not least of all) that kind of reason.

michaelpj commented:
Nah, it's fine. Mostly removing things, but we discussed that here and I'm fine with it. Just that that's the sort of thing that can end up bogging things down for ages.

kozross commented:
> Also, while I do like your size benchmarking suite, I still think it has to deal with the points I raised here: [#4250 (comment)](https://github.com/input-output-hk/plutus/issues/4250#issuecomment-981474284)
> 
> So in particular: does this make your actual validator scripts noticeably smaller? If you're pulling in the overhead elsewhere then it may not save you much.

I figured I would address this here rather than back in the issue for contextual reasons. While the points you raise _are_ valid, I still believe the work here is worthwhile. This is for two reasons:

* The on-chain size limit is really tight, and thus, any improvements, even shaving a few bytes, are worth it.
* Some of the reduction I demonstrate are so drastic that _not_ doing them seems like a huge missed opportunity.

With regard to the first point, we (MLabs) found that we run up against the on-chain size limit extremely quickly with any validators we want to write, especially when we use standard Haskell practices, including, but not limited to:

* Wrapping data types for correctness-by-construction
* Writing custom functions to deal with complex logic
* Factoring out common constructions in validator logic

In such settings, we have significantly many types and functions, all of which contribute. While this is in part our problem, I feel that the stuff provided by `plutus-tx` should be as minimal as possible: this is because any unnecessary contributions to script size from `plutus-tx` are difficult for application developers to address, while for blowup in their own functionality, this is substantially easier. Furthermore, even if the savings look tiny, they add up, since `Rational`s will likely be part of significantly more complex logic.

As for the second point, I think the following 'highlight reel' makes the point for my changes far better than my words ever could:

```
    On-chain size comparison for <:                   OK
      62B vs 112B 
      ~45% smaller
    On-chain size comparison for >:                   OK
      62B vs 112B 
      ~45% smaller
    On-chain size comparison for +:                   OK
      159B vs 222B 
      ~28% smaller
    On-chain size comparison for -:                   OK
      159B vs 222B 
      ~28% smaller
    On-chain size comparison for negate (overloaded): OK
      35B vs 326B (~0.5KiB) 
      ~89% smaller
    On-chain size comparison for %:                   OK
      163B vs 269B (~0.5KiB) 
      ~39% smaller
    On-chain size comparison for abs (overloaded):    OK
      69B vs 767B (~0.5KiB) 
      ~91% smaller
```

What these all have in common is that they are primitive operations: efficiency here benefits everyone, and these benefits are quite non-trivial. I think a lot of people would be quite surprised that use of `abs` eats over half a kibibyte of their on-chain size budget, without even taking the sizes of arguments into account! Even though the on-chain limit is 13KiB, given that argument sizes must be taken into consideration, this is realistically much closer to 4-5KiB in practice. Here is what these differences mean as 'fractions of your budget':

|Operations|Saving|Budget saved (assuming 5KiB)|
|--------------|---------|---------------------------------|
|``<``, ``>`` |50B    | ~1%                                |
|``+``, ``-``  |63B    | ~1.5%                             | 
|``%``         |106B  | ~2%                               |
|``negate`` |291B  | ~6%                                |
|``abs``      |698B  | ~14%                              |

Given how likely these are all to be used together (and in our experience, this has definitely been the case), this translates to significant savings. Thus, I believe that these are worthwhile savings to have: the constraints are so tight, and the benefits so significant, in enough cases that I feel it would be a shame not to.

michaelpj commented:
> While the points you raise are valid, I still believe the work here is worthwhile.

I don't think the work here is invalid. It seems useful. I just want to know what the impact actually _is_, and I don't believe that what you've shown me answers that question.

> As for the second point, I think the following 'highlight reel' makes the point for my changes far better than my words ever could:

No, because those benchmarks will massively overestimate the overhead, as I said in my other comment, because much of it will be shared. A simple way to see this would be to add a benchmark that uses all of the operations. I bet the reduction in size will be much more modest.

Again, I applaud the size benchmarking, I just think you're drawing much more drastic conclusions from it than you should, which is why I was curious about how much saving you get in an actual example.

kozross commented:
@michaelpj: As per your request, consider the function below:

```haskell
f :: Rational -> Rational -> Integer
f x y = round . negate $ cube (abs x) + cube (abs y)
   where
       cube :: Rational -> Rational
       cube z = z * z * z
```

I [implemented it](https://github.com/Liqwid-Labs/plutus-extra/blob/koz/sizing-proof/plutus-numeric/test/sizing-proof/Functions.hs#L33) using both the improved `Rational` (and attendant functions) I am PRing, as well as the ones currently in `plutus-tx`, [then compared the resulting function sizes](https://github.com/Liqwid-Labs/plutus-extra/blob/koz/sizing-proof/plutus-numeric/test/sizing-proof/Main.hs#L29). Relevant outcome is below:

```
  On-chain size comparison for round negated sum of absolute cubes (ours versus Plutus): OK
    578B (~0.5KiB) vs 1012B (~1KiB) 
    ~43% smaller
```

Does this adequately demonstrate the size difference you would like to see to be convinced? If not, could you be more specific about what precise criteria a demonstration would have to exhibit to convince you? I must admit that I'm not entirely sure what you're looking for, or by which criteria you are judging.

---

In somewhat other concerns, the CI is failing with an error that seems to have nothing to do with my changes. Is this expected?

kk-hainq commented:
> In somewhat other concerns, the CI is failing with an error that seems to have nothing to do with my changes. Is this expected?

Yes, that is a known flaky test case. Also, thank you for the effort!

michaelpj commented:
> Does this adequately demonstrate the size difference you would like to see to be convinced? If not, could you be more specific about what precise criteria a demonstration would have to exhibit to convince you?

Nice, that's much more encouraging! I would have expected the overhead reduction to be much less there, so this is good evidence that it's more useful than I expected :)

michaelpj commented:
Has genuine build failures in `plutus-tx-plugin` tests.

michaelpj commented:
Also conflicts.

michaelpj commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+  | n P.< P.zero = Rational (P.negate d) (P.negate n)
+  | P.True = Rational d n
+
+-- | Returns the whole-number part of its argument, dropping any leftover
```
Why not use the same description as `truncate` from the Haskell prelude? (This isn't wrong, just wondering)

kozross replied:
Because I find this clearer.

michaelpj commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+        | P.True         = gcd' b' (a' `Builtins.remainderInteger` b')
+
+-- | From GHC.Real
+-- | 'reduce' is a subsidiary function used only in this module.
```
No longer true. Maybe just say what it does.

michaelpj commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+        | b' P.== P.zero = a'
+        | P.True         = gcd' b' (a' `Builtins.remainderInteger` b')
+
+-- | From GHC.Real
```
```suggestion
-- From GHC.Real
```
otherwise I think Haddock will be unhappy

michaelpj reviewed (approved):
Some comments, but overall I'm happy.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+  {-# INLINEABLE one #-}
+  one = Rational P.one P.one
+
+instance P.ToData Rational where
```
They will very much not, especially for `unsafeFromBuiltinData` and `fromBuiltinData`. The generated instances will not check for normalization _or_ a zero denominator, as they use the constructor directly: my instances do both, with normalization handled by `%`.

michaelpj replied:
Good point. I don't know what to suggest about instances for types that have internal invariants. I guess this is okay, although it makes me uneasy.

michaelpj replied:
It needs a clear comment explaining why it is that way, at least.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+(%) :: Integer -> Integer -> Rational
+n % d
+  | d P.== P.zero = Builtins.error ()
+  | d P.< P.zero = P.negate n % P.negate d
```
Except I am not doing that? The constructor is deliberately left _un_exposed so that the invariants can be maintained without any risk of breakage. We even ensure this at serialization boundaries.

michaelpj replied:
Right, I can't read. Please document the invariant clearly, then.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+-- If given a zero denominator, this function will error. If you don't mind a
+-- size increase, and care about safety, use 'ratio' instead.
+{-# INLINEABLE (%) #-}
+(%) :: Integer -> Integer -> Rational
```
I tried where possible to keep to the API provided by `PlutusTx.Ratio` to minimize downstream breakage. In this case, `ratio` is the _safe_ one: are you suggesting that I rename `%`?

michaelpj replied:
Yes.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+  parseJSON = withObject "Rational" Haskell.$ \obj -> do
+    n <- obj .: "numerator"
+    d <- obj .: "denominator"
+    if d Haskell.== 0
```
Good point.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+
+-- | 0.5
+{-# INLINEABLE half #-}
+half :: Rational
```
Again, this is duplicating the original API. I wasn't keen on its inclusion, but don't want to break downstream code unnecessarily.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,65 +4,318 @@
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeMonoid Rational where
```
_Nothing_ currently has `Module` instances. I'd rather address this in a separate PR, since other instances can and should be added alongside.

michaelpj replied:
`Value` has one.

kozross replied:
Done.

kozross commented:
This is an interesting result:

```
reduce:                           FAIL
        ✗ reduce failed at test/StdLib/Spec.hs:104:5
          after 1 test.
        
              ┏━━ test/StdLib/Spec.hs ━━━
           96 ┃ testReduce :: Property
           97 ┃ testReduce = Hedgehog.property $ do
           98 ┃     let gen = Gen.integral (Range.linear (-10000) 100000)
           99 ┃     (n1, n2) <- Hedgehog.forAll $ (,) <$> gen <*> gen
              ┃     │ ( -10000 , -10000 )
          100 ┃     ghcResult <- tryHard $ reduce n1 n2
          101 ┃     plutusResult <- tryHard $ Ratio.toGHC $ Ratio.reduce n1 n2
          102 ┃     Hedgehog.annotateShow ghcResult
              ┃     │ Just ((-1) % (-1))
          103 ┃     Hedgehog.annotateShow plutusResult
              ┃     │ Just (1 % 1)
          104 ┃     Hedgehog.assert (ghcResult == plutusResult)
              ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        
          This failure can be reproduced by running:
          > recheck (Size 0) (Seed 5370228406319050580 5598176898127561611) reduce
        
      Use '--hedgehog-replay "Size 0 Seed 5370228406319050580 5598176898127561611"' to reproduce.
      
      Use -p '/reduce/' to rerun this test only.
```

This suggests that the `reduce` that we have is more eager, as it gives 1/1 instead of -1/-1. Given that this is the behaviour of `reduce` (which I have not altered from its original incarnation in terms of functionality), is this a bug in `reduce`, the test, or something else entirely?

@michaelpj - I assume the goal of this test was to ensure that we get coprimality between numerator and denominator, with the GHC `reduce` as an oracle. I can rewrite this test to do this, but this divergence isn't one I've added, and the current outcome is actually _more_ correct than the GHC one. What do you think is the right solution here?

kwxm commented on plutus-tx/src/PlutusTx/Numeric.hs:
```diff
@@ -126,3 +146,33 @@ instance MultiplicativeSemigroup a => Semigroup (Product a) where
+-- 'div' and 'mod' operations, which in turn are impemented as fst . divMod and
+-- snd . divMod, so we're calling Haskell's 'divMod' twice and throwing away
+-- useful information.  We could fix his by exposing 'divMod' as a PLC builtin,
+-- but that would require us to implement pairs as built-in types along with
```
Yes, that was written long before we had built-in pairs.

kwxm commented:
>  is this a bug in reduce, the test, or something else entirely?

I believe that what's going on here is that your version of `reduce` uses the `euclid` function, which is implemented using `mod`, whereas `reduce` (based on the [implementation](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.Real.html#reduce) in `GHC.Real`, which uses `[gcd](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.Real.html#gcd)`) implements Euclid's algorithm using `rem`, and `mod` and `rem` give different results when negative numbers are involved (see the lengthy note at the bottom of the file).  

The net result is that the sign of `euclid x y` is the same as the sign of `y`, but the sign of `gcd x y` is the same as the sign of `x`, and I think that accounts for the different reductions.  It's (very!) arguable what the "correct" result should, but for consistency I suppose that we should do the same as `GHC.Real` does, and we could presumably achieve that by using `Builtins.remainderInteger` instead of `Builtins.modInteger` in `euclid`.  The fact that Haskell has both `quot/rem` and `div/mod` is a perpetual source of annoyances like this.

kozross commented:
@kwxm Sorry for the late response - I was away for the holiday period. The very optimizations I use rest on the denominator always being positive: if we weaken this constraint, nothing I do matters. Furthermore, I would argue that the GHC implementation is wrong: -1 over -1 is 1 according to any reasonable mathematics, and having multiple representations for such a fundamental constant seems strange, when we don't accept multiple representations for, say, 2.

In terms of fixed-width integral types - @michaelpj has stated in the past that these are undesirable for _exactly_ reasons such as these, and I'm in agreement. Trying to future-proof an optimization for a choice that hasn't been made, and likely won't _ever_ be made, appears to be putting carts before horses.

In general, I would argue that this test just needs to be rewritten. GHC's implementations of anything aren't so great that we can't do better. Yes, it means we can't use GHC as a 'reference implementation' anymore, but that just means we need additional tests, [which I have already written anyway](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-numeric/test).

kwxm commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+unsafeRatio :: Integer -> Integer -> Rational
+unsafeRatio n d
+  | d P.== P.zero = Builtins.error ()
+  | d P.< P.zero = unsafeRatio (P.negate n) (P.negate d)
```
I think that `euclid a b` returns a number with the same absolute value as `gcd a b`, except that it's <=0 if b<0 and >= 0 if b>0. I think that means that we could miss out the second case because the third case would still do the right thing if `d < zero`.  That's rather subtle though, and if we were going to rely on that we'd need to be sure that it's correct (and have careful tests).

kozross replied:
This could potentially reduce code size, but in my tests, `unsafeRatio` is fairly small already. While I have [extensive test coverage](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-numeric/test) which could be ported to ensure this doesn't break, this is a fairly sizeable task. I'm happy to undertake it, but this PR is already nearly a month old.

kwxm replied:
Yes, I agree.  If there was going to be a big win in terms of code size it might be worth making that change, but it would make the code rather fragile. I don't have a strong opinion either for or against.

michaelpj replied:
If the code is going to live here, the tests should be here too! I'm okay with that happening later, though.

kwxm commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+
+-- Helpers
+
+-- Euclid's algorithm
```
This is a bit confusing because we have two versions of the Euclidean algortihm: `euclid` here (which uses `modInteger`) and `gcd'` in `gcd`, which uses `remainderInteger`.  Can we get away with just one of these?  They behave differently when negative numbers are involved and it's hard to work out what's going on.  GHC.Real uses something the same as `ghc`, but I have to admit that `euclid` seems a bit better behaved.

kwxm replied:
It also seems that `euclid` is always called with a positive second argument (in which case it always returns a positive result), except perhaps if we're scaling by a negative number in the `Module` instance, which we probably don't care too much about anyway.  That might be a useful thing to mention.

kozross replied:
I originally didn't expose `gcd`, but I was asked to put it back in by @michaelpj. I honestly agree that exposing this internal detail is unnecessary, and adds more code (and confusion!), and would prefer to have it gone.

michaelpj replied:
Computing `gcd` is independently useful in some cases, it's not *just* an implementation detail of `Rational`.

kwxm reviewed (approved):
I added some comments, but quite a lot of it's just quibbling about minor things in the comments.  Feel free to ignore that kind of thing.

kwxm commented:
> @kwxm Sorry for the late response - I was away for the holiday period.

Me too!

> The very optimizations I use rest on the denominator always being positive: if we weaken this constraint, nothing I do matters.

Yes, I think I'm coming round to your point of view now.  The thing that was worrying me was that the semantics of `reduce` are different from those of GHC's version, which might cause unexpected problems for users.  However, the source of GHC.Real says

> -- | 'reduce' is a subsidiary function used only in this module.

but then they export it anyway.  Maybe we could _not_ export it from this module (but still mention the different semantics) and get rid of the problematic test.  It seems unlikely that anyone would actually need to use `reduce`, but if we hide it we won't have to worry about that.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+  {-# INLINEABLE fromBuiltinData #-}
+  fromBuiltinData dat = case P.fromBuiltinData dat of
+    P.Nothing -> P.Nothing
+    P.Just (n, d) -> if d P.== P.zero
```
No, because that would `error`. I have to trap the zero-denominator case to give a `Nothing` instead.

kwxm replied:
Really?  It says
```
P.Just (n, d) -> if d P.== P.zero
                        then Builtins.error ()
                        else P.Just (unsafeRatio n d)
```
I thought that if it said
```
P.Just (n, d) ->  P.Just (unsafeRatio n d)
```
then that would do the same thing because `unsafeRatio` will call `error` in the zero case.  However, (a) I'm not sure how that would affect the compiled Plutus Core, and (b) it's probably better to be explicit about what's going on.  I don't think it does any harm doing it this way, so just ignore me.

michaelpj replied:
You're not returning `Nothing` on line 150, you're calling error! So I think Kenneth is right that the current code could be replaced by `unsafeRatio`, but perhaps you *meant* to return `Nothing` there?

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+  | P.True =
+    let gcd' = euclid n d
+     in P.Just P..
+        Rational (n `Builtins.quotientInteger` gcd') P.$
```
I prefer the `foo . bar $ baz` style of chaining arguments together most of the time. This one slipped by.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+-- = Note
+--
+-- This is specialized for 'Rational'; use this instead of the generic version
+-- of this function, as it is significantly smaller on-chain.
```
There used to be a function with the same name in this module (now in `PlutusTx.Numeric`), with the following signature: ``(AdditiveGroup a, Ord a) => a -> a``. Given that `Rational` is both an `AdditiveGroup` and `Ord`, that function can potentially be used with it. However, this is quite bad for size:

```
    On-chain size comparison for abs:                 OK
      644B (~0.5KiB) vs 767B (~0.5KiB) 
      ~16% smaller
    On-chain size comparison for abs (overloaded):    OK
      69B vs 767B (~0.5KiB) 
      ~91% smaller
```

While it would be nice to enable both efficiency _and_ a generic implementation, it's rather tricky, as the (more size-efficient) `abs` in this PR's version of `PlutusTx.Ratio` relies a lot on internal invariants of `Rational`. 

I will add an explanatory note to this effect.

kozross commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,349 @@
+-- * @n@ and @f@ both have the same sign as @r@; and
+-- * @'abs' f 'P.<' 'P.one'@.
+{-# INLINEABLE properFraction #-}
+properFraction :: Rational -> (Integer, Rational)
```
This is a slight rewording of the documentation for 'properFraction' in [the corresponding GHC module](https://hackage.haskell.org/package/base-4.16.0.0/docs/GHC-Real.html#v:properFraction), modulo detail not relevant to this case (such as the statement that `n` is an integral). I'm happy to extend this, but I would rather have a simpler description that one which pedantically spells out how all invariants are preserved. Ultimately, users shouldn't need to know our implementation internals.

kozross commented:
@kwxm FWIW, I've repaired the offending test by adding a normalization step. It's a little annoying to do, but it retains most of the original test and doesn't seem to cause any issues. I've also responded to all of your comments, some with changes to the PR.

@michaelpj Can we merge this? I believe I've addressed all concerns.

michaelpj commented on plutus-tx/src/PlutusTx/Ratio.hs:
```diff
@@ -4,80 +4,362 @@
+instance P.UnsafeFromData Rational where
+  {-# INLINABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData dat = case P.unsafeFromBuiltinData dat of
+    (n, d) -> if d P.== P.zero
```
This one I think really could just be `unsafeRatio`.

kozross replied:
That's a very good catch - thank you for sticking to me about this! As a side effect, the new version of `fromBuiltinData` has an on-chain size that's 8 bytes smaller: not much, but every little helps.

michaelpj reviewed (commented):
Sorry, a couple more small things then this is good to go.

kozross commented:
@michaelpj Addressed your concerns. Where test porting is concerned, I would like to do this in a separate PR; additionally, feature request #4314 directly concerns such a PR, specifically regarding size testing.

michaelpj commented:
Thanks!
